### ISSUE TITLE
Fails to build with clang-14

### ISSUE THREAD

**Issue by yurivict** [Qiskit/qiskit-aer#1573]
### Informations

- **Qiskit Aer version**: 0.10.4
- **Python version**: 3.9
- **Operating system**: FreeBSD 13.1

### What is the current behavior?
```
===>  Building for qiskit-aer-0.10.4
[100% 1/1] : && /usr/bin/c++ -O2 -pipe -fno-omit-frame-pointer -fstack-protector-strong -isystem /usr/local/include -fno-strict-aliasing -fno-omit-frame-pointer  -isystem /usr/local/include -pedantic -Wall -Wfloat-equal -Wundef -Wcast-align -Wwrite-strings -Wmissing-declarations -Wredundant-decls -Wshadow -Woverloaded-virtual -fopenmp=libomp -O2 -pipe -fno-omit-frame-pointer -fstack-protector-strong -isystem /usr/local/include -fno-strict-aliasing -fno-omit-frame-pointer  -isystem /usr/local/include -fstack-protector-strong -L/usr/local/lib       -fopenmp=libomp CMakeFiles/qasm_simulator.dir/contrib/standalone/qasm_simulator.cpp.o -o Release/qasm_simulator  /usr/local/lib/libopenblas.so  -lspdlog  -pthread && :
FAILED: Release/qasm_simulator 
: && /usr/bin/c++ -O2 -pipe -fno-omit-frame-pointer -fstack-protector-strong -isystem /usr/local/include -fno-strict-aliasing -fno-omit-frame-pointer  -isystem /usr/local/include -pedantic -Wall -Wfloat-equal -Wundef -Wcast-align -Wwrite-strings -Wmissing-declarations -Wredundant-decls -Wshadow -Woverloaded-virtual -fopenmp=libomp -O2 -pipe -fno-omit-frame-pointer -fstack-protector-strong -isystem /usr/local/include -fno-strict-aliasing -fno-omit-frame-pointer  -isystem /usr/local/include -fstack-protector-strong -L/usr/local/lib       -fopenmp=libomp CMakeFiles/qasm_simulator.dir/contrib/standalone/qasm_simulator.cpp.o -o Release/qasm_simulator  /usr/local/lib/libopenblas.so  -lspdlog  -pthread && :
ld: error: undefined symbol: AER::QV::Avx AER::QV::apply_matrix_avx<double>(double*, unsigned long, unsigned long const*, unsigned long, double const*, unsigned long)
>>> referenced by qasm_simulator.cpp
>>>               CMakeFiles/qasm_simulator.dir/contrib/standalone/qasm_simulator.cpp.o:(AER::QV::TransformerAVX2<std::__1::complex<double>*, double>::apply_matrix(std::__1::complex<double>*&, unsigned long, int, std::__1::vector<unsigned long, std::__1::allocator<unsigned long> > const&, std::__1::vector<std::__1::complex<double>, std::__1::allocator<std::__1::complex<double> > > const&) const)

ld: error: undefined symbol: AER::QV::Avx AER::QV::apply_diagonal_matrix_avx<double>(double*, unsigned long, unsigned long const*, unsigned long, double const*, unsigned long)
>>> referenced by qasm_simulator.cpp
>>>               CMakeFiles/qasm_simulator.dir/contrib/standalone/qasm_simulator.cpp.o:(AER::QV::TransformerAVX2<std::__1::complex<double>*, double>::apply_diagonal_matrix(std::__1::complex<double>*&, unsigned long, int, std::__1::vector<unsigned long, std::__1::allocator<unsigned long> > const&, std::__1::vector<std::__1::complex<double>, std::__1::allocator<std::__1::complex<double> > > const&) const)

ld: error: undefined symbol: AER::QV::Avx AER::QV::apply_matrix_avx<float>(float*, unsigned long, unsigned long const*, unsigned long, float const*, unsigned long)
>>> referenced by qasm_simulator.cpp
>>>               CMakeFiles/qasm_simulator.dir/contrib/standalone/qasm_simulator.cpp.o:(AER::QV::TransformerAVX2<std::__1::complex<float>*, float>::apply_matrix(std::__1::complex<float>*&, unsigned long, int, std::__1::vector<unsigned long, std::__1::allocator<unsigned long> > const&, std::__1::vector<std::__1::complex<double>, std::__1::allocator<std::__1::complex<double> > > const&) const)

ld: error: undefined symbol: AER::QV::Avx AER::QV::apply_diagonal_matrix_avx<float>(float*, unsigned long, unsigned long const*, unsigned long, float const*, unsigned long)
>>> referenced by qasm_simulator.cpp
>>>               CMakeFiles/qasm_simulator.dir/contrib/standalone/qasm_simulator.cpp.o:(AER::QV::TransformerAVX2<std::__1::complex<float>*, float>::apply_diagonal_matrix(std::__1::complex<float>*&, unsigned long, int, std::__1::vector<unsigned long, std::__1::allocator<unsigned long> > const&, std::__1::vector<std::__1::complex<double>, std::__1::allocator<std::__1::complex<double> > > const&) const)
c++: error: linker command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.
===> Compilation failed unexpectedly.
Try to set MAKE_JOBS_UNSAFE=yes and rebuild before reporting the failure to
the maintainer.
*** Error code 1
```

### Steps to reproduce the problem
normal build

### What is the expected behavior?
n/a

### Suggested solutions
Please test build with clang.

**Comment by yurivict** [comment]
gcc-11 also fails to compile it:
```
===>  Building for qiskit-aer-0.10.4
[100% 1/1] : && /usr/local/bin/g++11 -O2 -pipe -fno-omit-frame-pointer  -fstack-protector-strong -Wl,-rpath=/usr/local/lib/gcc11 -fno-omit-frame-pointer  -Wl,-rpath=/usr/local/lib/gcc11 -isystem /usr/local/include -pedantic -Wall -Wfloat-equal -Wundef -Wcast-align -Wwrite-strings -Wmissing-declarations -Wredundant-decls -Wshadow -Woverloaded-virtual -fopenmp -O2 -pipe -fno-omit-frame-pointer  -fstack-protector-strong -Wl,-rpath=/usr/local/lib/gcc11 -fno-omit-frame-pointer  -Wl,-rpath=/usr/local/lib/gcc11 -isystem /usr/local/include -fstack-protector-strong -Wl,-rpath=/usr/local/lib/gcc11 -L/usr/local/lib/gcc11 -L/usr/local/lib       -fopenmp CMakeFiles/qasm_simulator.dir/contrib/standalone/qasm_simulator.cpp.o -o Release/qasm_simulator  /usr/local/lib/libopenblas.so  -lspdlog  -pthread && :
FAILED: Release/qasm_simulator 
: && /usr/local/bin/g++11 -O2 -pipe -fno-omit-frame-pointer  -fstack-protector-strong -Wl,-rpath=/usr/local/lib/gcc11 -fno-omit-frame-pointer  -Wl,-rpath=/usr/local/lib/gcc11 -isystem /usr/local/include -pedantic -Wall -Wfloat-equal -Wundef -Wcast-align -Wwrite-strings -Wmissing-declarations -Wredundant-decls -Wshadow -Woverloaded-virtual -fopenmp -O2 -pipe -fno-omit-frame-pointer  -fstack-protector-strong -Wl,-rpath=/usr/local/lib/gcc11 -fno-omit-frame-pointer  -Wl,-rpath=/usr/local/lib/gcc11 -isystem /usr/local/include -fstack-protector-strong -Wl,-rpath=/usr/local/lib/gcc11 -L/usr/local/lib/gcc11 -L/usr/local/lib       -fopenmp CMakeFiles/qasm_simulator.dir/contrib/standalone/qasm_simulator.cpp.o -o Release/qasm_simulator  /usr/local/lib/libopenblas.so  -lspdlog  -pthread && :
/usr/local/bin/ld: CMakeFiles/qasm_simulator.dir/contrib/standalone/qasm_simulator.cpp.o: in function `AER::QV::TransformerAVX2<std::complex<double>*, double>::apply_diagonal_matrix(std::complex<double>*&, unsigned long, int, std::vector<unsigned long, std::allocator<unsigned long> > const&, std::vector<std::complex<double>, std::allocator<std::complex<double> > > const&) const':
qasm_simulator.cpp:(.text._ZNK3AER2QV15TransformerAVX2IPSt7complexIdEdE21apply_diagonal_matrixERS4_miRKSt6vectorImSaImEERKS7_IS3_SaIS3_EE[_ZNK3AER2QV15TransformerAVX2IPSt7complexIdEdE21apply_diagonal_matrixERS4_miRKSt6vectorImSaImEERKS7_IS3_SaIS3_EE]+0x5a): undefined reference to `AER::QV::Avx AER::QV::apply_diagonal_matrix_avx<double>(double*, unsigned long, unsigned long const*, unsigned long, double const*, unsigned long)'
/usr/local/bin/ld: CMakeFiles/qasm_simulator.dir/contrib/standalone/qasm_simulator.cpp.o: in function `AER::QV::TransformerAVX2<std::complex<float>*, float>::apply_diagonal_matrix(std::complex<float>*&, unsigned long, int, std::vector<unsigned long, std::allocator<unsigned long> > const&, std::vector<std::complex<double>, std::allocator<std::complex<double> > > const&) const':
qasm_simulator.cpp:(.text._ZNK3AER2QV15TransformerAVX2IPSt7complexIfEfE21apply_diagonal_matrixERS4_miRKSt6vectorImSaImEERKS7_IS2_IdESaISC_EE[_ZNK3AER2QV15TransformerAVX2IPSt7complexIfEfE21apply_diagonal_matrixERS4_miRKSt6vectorImSaImEERKS7_IS2_IdESaISC_EE]+0x5a): undefined reference to `AER::QV::Avx AER::QV::apply_diagonal_matrix_avx<float>(float*, unsigned long, unsigned long const*, unsigned long, float const*, unsigned long)'
/usr/local/bin/ld: CMakeFiles/qasm_simulator.dir/contrib/standalone/qasm_simulator.cpp.o: in function `AER::QV::TransformerAVX2<std::complex<float>*, float>::apply_matrix(std::complex<float>*&, unsigned long, int, std::vector<unsigned long, std::allocator<unsigned long> > const&, std::vector<std::complex<double>, std::allocator<std::complex<double> > > const&) const':
qasm_simulator.cpp:(.text._ZNK3AER2QV15TransformerAVX2IPSt7complexIfEfE12apply_matrixERS4_miRKSt6vectorImSaImEERKS7_IS2_IdESaISC_EE[_ZNK3AER2QV15TransformerAVX2IPSt7complexIfEfE12apply_matrixERS4_miRKSt6vectorImSaImEERKS7_IS2_IdESaISC_EE]+0xe9): undefined reference to `AER::QV::Avx AER::QV::apply_matrix_avx<float>(float*, unsigned long, unsigned long const*, unsigned long, float const*, unsigned long)'
/usr/local/bin/ld: CMakeFiles/qasm_simulator.dir/contrib/standalone/qasm_simulator.cpp.o: in function `AER::QV::TransformerAVX2<std::complex<double>*, double>::apply_matrix(std::complex<double>*&, unsigned long, int, std::vector<unsigned long, std::allocator<unsigned long> > const&, std::vector<std::complex<double>, std::allocator<std::complex<double> > > const&) const':
qasm_simulator.cpp:(.text._ZNK3AER2QV15TransformerAVX2IPSt7complexIdEdE12apply_matrixERS4_miRKSt6vectorImSaImEERKS7_IS3_SaIS3_EE[_ZNK3AER2QV15TransformerAVX2IPSt7complexIdEdE12apply_matrixERS4_miRKSt6vectorImSaImEERKS7_IS3_SaIS3_EE]+0xe9): undefined reference to `AER::QV::Avx AER::QV::apply_matrix_avx<double>(double*, unsigned long, unsigned long const*, unsigned long, double const*, unsigned long)'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
===> Compilation failed unexpectedly.
```

**Comment by yurivict** [comment]
This PR solves this problem: https://github.com/Qiskit/qiskit-aer/pull/1576

**Comment by jakelishman** [comment]
Fixed by #1576.
